src/pools/dynamic-choose-worker.ts

```diff
@@ -0,0 +1,42 @@
+import type { Worker as ClusterWorker } from 'cluster'
+import type { JSONValue } from '../utility-types'
+import type { DynamicClusterPool } from './cluster/dynamic'
+import type { DynamicThreadPool } from './thread/dynamic'
+import type { ThreadWorkerWithMessageChannel } from './thread/fixed'
+
+export function dynamicallyChooseWorker<
```

Comment:
I think an object with an interface shall be created for that function.
Basically, it's an implementation possible for the worker choice strategy and to follow the strategy design pattern, we have to be able to set/register an instance of the strategy at pool creation. 
We can be less strict regarding the design pattern respect: register callback following an interface or even function signature, but the global context initialisation idea shall be kept.

Reply:
I also thought about something like that, but currently the member functions have `protected` access
Using an interface for these will force to make them `public`
And I don't know if we want that :slightly_frowning_face: 

On the other side: providing different types of `chooseWorker` algos will help and improve our library
But I would prefer to provide this in a later version like v2.1+

---

So I have some minor improvements made in this branch, like the finding of a free worker section.
Maybe we should only merge https://github.com/pioardi/poolifier/pull/152/commits/858af07187ce65ef238468232dba898400cf054c and https://github.com/pioardi/poolifier/pull/152/commits/cd9cf243a1681fa0808d68015934f9dc172efbba but not the last commit with the extraction to `dynamicallyChooseWorker`

Reply:
We can start to sketch the requirement for a: 
+ FIFO algo
+ RR algo

FIFO needs more than a nextWorkerIndex reference, that means that enforcing data structure inside the abstract class for such strategy will not fit. 
FIFO needs a stack of free workers (by definition):
+ worker data structure can have a flag stating its status and last used timestamp: lookup for free worker is somewhat heavy inside the worker array ... 
+ pool have a customisable data structure that can be defined at strategy registration, for FIFO, it can be a simple array (push/pop), for RR it's just the next worker index

LIFO algo will also have similar requirement. 
Fair share algo also: worker usage time series and maths on them.

Reply:
> So I have some minor improvements made in this branch, like the finding of a free worker section.
> Maybe we should only merge [858af07](https://github.com/pioardi/poolifier/commit/858af07187ce65ef238468232dba898400cf054c) and [cd9cf24](https://github.com/pioardi/poolifier/commit/cd9cf243a1681fa0808d68015934f9dc172efbba) but not the last commit with the extraction to `dynamicallyChooseWorker`

Yes, please merge the improvements made into master branch and let the discussion goes on the factorisation and hopefully settles before merging it.